libp2p/crypto/minasn1.nim

```diff
@@ -406,54 +412,52 @@ proc asn1EncodeContextTag*(dest: var openarray[byte], value: openarray[byte],
     copyMem(addr dest[1], addr buffer[0], lenlen)
     copyMem(addr dest[1 + lenlen], unsafeAddr value[0], len(value))
 
-proc getLength(ab: var Asn1Buffer, length: var uint64): Asn1Status =
+proc getLength(ab: var Asn1Buffer): Asn1Result[uint64] =
   ## Decode length part of ASN.1 TLV triplet.
-  result = Asn1Status.Incomplete
   if not ab.isEmpty():
     let b = ab.buffer[ab.offset]
     if (b and 0x80'u8) == 0x00'u8:
-      length = cast[uint64](b)
+      let length = cast[uint64](b)
       ab.offset += 1
-      result = Asn1Status.Success
-      return
+      return ok(length)
     if b == 0x80'u8:
-      length = 0'u64
-      result = Asn1Status.Indefinite
-      return
+      return err(Asn1Error.Indefinite)
     if b == 0xFF'u8:
-      length = 0'u64
-      result = Asn1Status.Incorrect
-      return
+      return err(Asn1Error.Incorrect)
     let octets = cast[uint64](b and 0x7F'u8)
     if octets > 8'u64:
-      length = 0'u64
-      result = Asn1Status.Overflow
-      return
-    length = 0'u64
+      return err(Asn1Error.Overflow)
     if ab.isEnough(int(octets)):
+      var length: uint64 = 0
       for i in 0..<int(octets):
         length = (length shl 8) or cast[uint64](ab.buffer[ab.offset + i + 1])
       ab.offset = ab.offset + int(octets) + 1
-      result = Asn1Status.Success
+      return ok(length)
+    else:
+      return err(Asn1Error.Incomplete)
+  else:
+    return err(Asn1Error.Incomplete)
 
-proc getTag(ab: var Asn1Buffer, tag: var int,
-            klass: var Asn1Class): Asn1Status =
+proc getTag(ab: var Asn1Buffer, tag: var int): Asn1Result[Asn1Class] =
   ## Decode tag part of ASN.1 TLV triplet.
-  result = Asn1Status.Incomplete
   if not ab.isEmpty():
     let b = ab.buffer[ab.offset]
     var c = int((b and 0xC0'u8) shr 6)
+    var klass: Asn1Class
     if c >= 0 and c < 4:
       klass = cast[Asn1Class](c)
     else:
-      return Asn1Status.Incorrect
+      return err(Asn1Error.Incorrect)
     tag = int(b and 0x3F)
     ab.offset += 1
-    result = Asn1Status.Success
+    return ok(klass)
```

Comment:
no return needed here

Reply:
An explicit "return" is more readable.

BTW, I wonder if it's possible to add new lines by mistake after such a magical Ruby/Lisp-like result-line. The types would have to match.

Reply:
if you remove the return, the compiler ensures that all branches have a value/expression at the end of matching type - that removes some possibilities for a class of mistakes

Reply:
basically, using return increases the risk of some branches not returning anything, specially since nim allows a branch _not_ to have a return as soon as one branch has a return - together with indent-based scope, it's an unnecessary risk

Reply:
Sounds like code obfuscation as a workaround for a compiler bug.

Reply:
compiler bug? the language is designed in such a way that you don't have to `return` at the end of a function or statement. 

here's an example of a perfectly valid nim function that silently falls through:
```
proc f(): int =
  if false:
    return 42
```

this is a ridiculous example - but nest that if 2-3 levels as is typical for nim code and you're looking at a whole new level of "unreadable".

you can avoid this issue by relying on expressions - `if`, `try` etc, as long as they end is a value, are expressions.

```
proc f(): int =
  if false:
    42  # compiler error - the 42 has nowhere to go now - it's no longer "consumed" by `return` or `discard` so you have to do something with it - the magic of expressions!
```

Reply:
It won't be "42". It will be "foo(bar)" and you're supposed to guess it returns something and that that something is at the end of some control flow branch so it gets automatically returned by the enclosing procedure.

Reply:
well, that's the point kind of, of turning statements into expressions: you no longer has to guess because the compiler checks it for you - the guessing happens when you use `return` and statements in general: did all branches have a value? you can only guess (in a longer example) and hope, because the compiler doesn't check it for you.

this is not limited to functions:
```
var v: int
if x:
  v= 4 # compiler doesn't check that v was assigned in all branches

let v =
  if x:
    4 # compile does check that v gets a value in all branches
```